src/urllib3/response.py

```diff
@@ -785,12 +763,86 @@ def read(
             self._fp_bytes_read += len(data)
             if self.length_remaining is not None:
                 self.length_remaining -= len(data)
+        return data
+
+    def _popleft_decoded_bytes(self, n: int) -> bytes:
+        decoded_data = bytes(self._decoded_bytes[:n])
+        del self._decoded_bytes[:n]
+        return decoded_data
+
+    def _popall_decoded_bytes(self) -> bytes:
+        decoded_data = bytes(self._decoded_bytes)
+        del self._decoded_bytes[:]
+        return decoded_data
 
-            data = self._decode(data, decode_content, flush_decoder)
+    def _decoded_bytes_in_buffer(self) -> int:
+        return len(self._decoded_bytes)
 
+    def read(
+        self,
+        amt: Optional[int] = None,
+        decode_content: Optional[bool] = None,
+        cache_content: bool = False,
+    ) -> bytes:
+        """
+        Similar to :meth:`http.client.HTTPResponse.read`, but with two additional
+        parameters: ``decode_content`` and ``cache_content``.
+
+        :param amt:
+            How much of the content to read. If specified, caching is skipped
+            because it doesn't make sense to cache partial content as the full
+            response.
+
+        :param decode_content:
+            If True, will attempt to decode the body based on the
+            'content-encoding' header.
+
+        :param cache_content:
+            If True, will save the returned data such that the same result is
+            returned despite of the state of the underlying file object. This
+            is useful if you want the ``.data`` property to continue working
+            after having ``.read()`` the file object. (Overridden if ``amt`` is
+            set.)
+        """
+        if amt is not None:
+            cache_content = False
+
+        if decode_content is None:
+            decode_content = self.decode_content
+
+        if not decode_content:
+            data = self.raw_read(amt)
             if cache_content:
                 self._body = data
+            return data
+
+        if amt is not None and self._decoded_bytes_in_buffer() >= amt:
+            return self._popleft_decoded_bytes(amt)
+
+        data = self.raw_read(amt)
+
+        if not data and self._decoded_bytes_in_buffer() == 0:
+            return data
+
+        self._init_decoder()
```

Comment:
Would splitting all the logic for amt is None and amt is not None here into their own blocks simplify both implementations?

Reply:
Definitely: [`57c5778` (#2712)](https://github.com/urllib3/urllib3/pull/2712/commits/57c5778712209566b8743167520510b61a28f848)